### PR TITLE
update shape to `ivy.Shape` and `ivy.NativeShape` 

### PR DESCRIPTION
**Pull request by juliagsy**
- added `ivy.Shape` and `ivy.NativeShape` classes
- updated type hints and returns in ivy and backends
- added shape_array_mode methods (set, unset) with docstrings
- implemented array wrapping to shapes

todo: will update `shape_to_tuple` method to the equivalent shape and native shape methods in another PR!

thanks!

### CODE REVIEW

File: ivy/__init__.py
Comment (djl11) [inline ivy/__init__.py:69]: you can replace `ivy.current_backend().NativeShape` with `ivy.NativeShape`, if a backend is currently set, it will have overwritten `ivy.NativeShape`

File: ivy/__init__.py
Comment (djl11) [inline ivy/__init__.py:72]: small detail, but there is no need to cast to `list` first. You can just do `shape_tup = (shape_tup,)`

File: ivy/__init__.py
Comment (djl11) [inline ivy/__init__.py:76]: `assert all([isinstance(v, int) for v in shape_tup])`

File: ivy/functional/backends/numpy/random.py
Comment (djl11) [inline ivy/functional/backends/numpy/random.py:7]: typo

File: ivy/functional/ivy/general.py
Comment (djl11) [inline ivy/functional/ivy/general.py:2566]: this should use a stack rather than a single global value enabling us to use nested blocks of setting and unsetting. See other global variables for examples.

**Review by djl11 — CHANGES_REQUESTED** [review]
Please see comments :)

Didn't go through in crazy detail. Once you've done, request another review, and provided that no new tests are failing, we can get this merged pretty quickly. Thanks!

File: ivy/__init__.py
Comment (djl11) [inline ivy/__init__.py:73]: no need to use `ivy.all` and `ivy.array` here, just `all([isinstance(v, int) for v in shape_tup])`

**Review by djl11 — CHANGES_REQUESTED** [review]
small, change requested, then feel free to merge this yourself provided the tests are passing!

File: ivy/__init__.py
Comment (juliagsy) [reply ivy/__init__.py:73]: some of the functions threw array type errors when I used them directly 😅

or let's say we merge it first and I'll find a fix in another pr since I'll be changing up the `shape_to_tuple` methods in a new PR after merging this?

**Pull request merged**
